### ISSUE TITLE
c.Env[*] values lost in the subrouter mapping

### ISSUE THREAD

**Issue by jeevatkm** [zenazn/goji#150]
I may be missing somethings, need help here! My setup is quite similar to [example](https://github.com/zenazn/goji/blob/master/example/main.go).

<pre><code>// Applying middleware's here, typically manipulates and then adds object or value to c.Env[---]
goji.Use(ApplyTemplates)
goji.Use(ApplySessions)
goji.Use(ApplyDatabase)
goji.Use(ApplyAuth) // checks the session user id and populates user into c.Env["User"]

admin := web.New()
goji.Handle("/admin/*", admin)
admin.Use(middleware.SubRouter)
admin.Use(ApplyAdminAuth) // Retrieving c.Env["User"] and need check for Admin Permission
admin.Get("/", Dashboard)
goji.Get("/admin", http.RedirectHandler("/admin/", 301))
</code></pre>

Facing issue at 
<pre>admin.Use(ApplyAdminAuth) // Retrieving c.Env["User"] and need check for Admin Permission</pre>

I'm getting <code>nil</code> for <code>c.Env["User"]</code> in the <code>ApplyAdminAuth</code> middleware. It seems values are not made it sub path <code>/admin/*</code>.

Where I'm making a mistake?

**Comment by elithrar** [comment]
Are ApplyAuth and ApplyAdminAuth potentially conflicting? Remember that all
parent router middleware is applied to the sub router. If you can post the
code for those two middleware it'll help.
You can also sprinkle log.Printf("%v", c.Env) along the way to dump the
contents of the request context to logs as well, which will help debug.

**Comment by jeevatkm** [comment]
Yes, that's my understanding too. Parent middleware applies to sub router.

Middleware code:
<pre>// This is parent one
func ApplyAuth(c *web.C, h http.Handler) http.Handler {
	fn := func(w http.ResponseWriter, r *http.Request) {
		session := c.Env["Session"].(*sessions.Session)
		if userId, ok := session.Values["User"].(bson.ObjectId); ok {
			dbSession := c.Env["DBSession"].(*mgo.Session)
			database := dbSession.DB(c.Env["DBName"].(string))
			log.Println("I'm inside user check")

			user := new(model.User)
			err := database.C("users").Find(bson.M{"_id": userId}).One(&user)
			if err != nil {
				log.Printf("Auth error: %v", err)
				c.Env["User"] = nil
			} else {
				c.Env["User"] = user
			}
		}
		h.ServeHTTP(w, r)
	}
	return http.HandlerFunc(fn)
}

// This one applies to sub router /admin/*
func ApplyAdminAuth(c *web.C, h http.Handler) http.Handler {
	fn := func(w http.ResponseWriter, r *http.Request) {
		log.Printf("/admin/* ==> %s", r.URL.RequestURI())
		log.Printf("ApplyAdminAuth User Info: %v", c.Env["User"])
		h.ServeHTTP(w, r)
	}
	return http.HandlerFunc(fn)
}</pre>

**Comment by jeevatkm** [comment]
Any pointers would be helpful? I'm still analyzing it no luck!

**Comment by zenazn** [comment]
That's super weird! I wrote a little testing app which seems to work:

```
package main

import (
	"fmt"
	"log"
	"net/http"

	"github.com/zenazn/goji"
	"github.com/zenazn/goji/web"
	"github.com/zenazn/goji/web/middleware"
)

func Outer(c *web.C, h http.Handler) http.Handler {
	fn := func(w http.ResponseWriter, r *http.Request) {
		c.Env = map[interface{}]interface{}{
			"user": "carl",
		}
		h.ServeHTTP(w, r)
	}
	return http.HandlerFunc(fn)
}

func Inner(c *web.C, h http.Handler) http.Handler {
	fn := func(w http.ResponseWriter, r *http.Request) {
		log.Print(c.Env["user"])
		h.ServeHTTP(w, r)
	}
	return http.HandlerFunc(fn)
}

func main() {
	goji.Use(Outer)
	sub := web.New()
	goji.Get("/sub/*", sub)
	sub.Use(middleware.SubRouter)
	sub.Use(Inner)
	sub.Get("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "hello world")
	})

	goji.Serve()
}
```

I'm not sure why this would work but your app wouldn't. Would you be able to let me look at more of the surrounding code / middleware?

**Comment by jeevatkm** [comment]
Thanks @zenazn and @elithrar for your inputs. Above sample code works fine. It seems I'm doing something wrong! 

I will give a one more try! if not success then I will share my code base with you.

**Comment by jeevatkm** [comment]
Okay, finally my journey ended with smile :)

Week ago I was started my application based on idea from this [app](https://github.com/elcct/defaultproject) and improved so on. Sample project gave an idea for Go-Lang and good starting point. Since I came from Java and Objective-C background.

However I ended up with issue mentioned in this thread and also few other issues as well. Will share my  experience to the project author later. 

After @zenazn sample code, I confirmed myself I'm at wrong track. Finally I have restructured my application based on [example](https://github.com/zenazn/goji/blob/master/example/main.go) and [this article](http://elithrar.github.io/article/custom-handlers-avoiding-globals/).  Able to over come issues I faced.
